vm/devices/storage/disklayer_sqlite/src/lib.rs

```diff
@@ -0,0 +1,617 @@
+            .map_err(|e| DiskError::Io(std::io::Error::new(std::io::ErrorKind::Other, e)))?;
+
+        Ok(())
+    }
+
+    async fn unmap(
+        &self,
+        sector_offset: u64,
+        sector_count: u64,
+        _block_level_only: bool,
+        next_is_zero: bool,
+    ) -> Result<(), DiskError> {
+        tracing::trace!(sector_offset, sector_count, "unmap");
+        if sector_offset + sector_count > self.meta.sector_count {
+            return Err(DiskError::IllegalBlock);
+        }
+
+        unblock({
+            let conn = self.conn.clone();
+            move || unmap_sectors(conn, sector_offset, sector_count, next_is_zero)
+        })
+        .await
+        .map_err(|e| DiskError::Io(std::io::Error::new(std::io::ErrorKind::Other, e)))?;
+
+        Ok(())
+    }
+
+    fn unmap_behavior(&self) -> UnmapBehavior {
+        // This layer zeroes if the lower layer is zero, but otherwise does
+        // nothing, so we must report unspecified.
+        UnmapBehavior::Unspecified
+    }
+
+    fn optimal_unmap_sectors(&self) -> u32 {
+        1
+    }
+}
+
+impl WriteNoOverwrite for SqliteDiskLayer {
+    async fn write_no_overwrite(
+        &self,
+        buffers: &RequestBuffers<'_>,
+        sector: u64,
+    ) -> Result<(), DiskError> {
+        self.write_maybe_overwrite(buffers, sector, false).await
+    }
+}
+
+#[allow(clippy::large_enum_variant)]
+enum SectorKind {
+    AllZero,
+    AllOne,
+    Data(Vec<u8>),
+}
+
+// FUTURE: pass RequestBuffers into this function directly, and avoid the alloc
+// overhead.
+fn read_sectors(
+    conn: Arc<Mutex<Connection>>,
+    sector_size: u32,
+    start_sector: u64,
+    end_sector: u64,
+) -> Result<Vec<(u64, SectorKind)>, rusqlite::Error> {
+    let conn = conn.lock();
+
+    let mut select_stmt = conn.prepare_cached(
+        "SELECT sector, data
+        FROM sectors
+        WHERE sector >= ? AND sector < ?
+        ORDER BY sector ASC",
+    )?;
+    let mut rows = select_stmt.query(rusqlite::params![start_sector, end_sector])?;
+
+    // reuse the same blob handle across all reads.
+    //
+    // this handle isn't easy to persist across multiple calls to
+    // `read_sectors`, due to the lifetime restrictions imposed by using a
+    // `Arc<Mutex<Connection>>`.
+    let Some(first_row) = rows.next()? else {
+        return Ok(Vec::new());
+    };
+    let mut sector: u64 = first_row.get(0)?;
+    let mut blob = conn.blob_open(
+        rusqlite::DatabaseName::Main,
+        "sectors",
+        "data",
+        sector as i64,
+        true,
+    )?;
+
+    let mut buf = vec![0; sector_size as usize];
+    let mut res = Vec::new();
+    loop {
+        let data = match blob.len() {
+            0 => SectorKind::AllZero,
+            1 => SectorKind::AllOne,
+            len => {
+                if len != sector_size as usize {
+                    return Err(rusqlite::Error::BlobSizeError);
+                }
+                blob.read_at(&mut buf, 0)?;
```

Comment:
It seems like this could fail with SQLITE_ABORT if there's a concurrent update to the sector: <https://www.sqlite.org/c3ref/blob_open.html>

Reply:
(It is not obvious to me that going through this blob interface is better than just getting the results via a normal select.)

Reply:
As written, you're correct that we're not really getting anything with going through the BLOB interface.

That said - I do think that we'll eventually want to lean on the blob interface in order to avoid allocations on read/write paths, both externally, with these Vecs that are getting allocated on the Rust side, but also internally, due to the use of SQLITE_TRANSIENT - see https://github.com/rusqlite/rusqlite/issues/164.

That said... it might be easier to just stick to the naive API and sidestep the SQLITE_ABORT issues, and then think about how to solve that using the blob API at some later point in time.

Reply:
The SQLITE_TRANSIENT issue seems to apply only to the binding code, which will only affect us for updates/inserts. Hopefully they'll just fix this sooner or later (although I think it will probably be a breaking change).

It looks like the `ValueRef` machinery will already get us minimal copies on the query side (once we figure out how to send the RequestBuffers over to the unblock'd task). I don't see where sqlite or rusqlite is allocating there.

Reply:
(Just make sure you use `.get_ref()` to get the column from the row)

Reply:
Oh, interesting. I didn't realize that the queries were already in good shape. Cool.